### PR TITLE
Function: Add FB Login + Views

### PR DESCRIPTION
**Pull request by krnjn**

### CODE REVIEW

**Comment by krnjn** [comment]
@stedmanblake some basic context of what I did here:

**Changes we def want to make**
- Moved all config variables / app ids / keys to ENV vars – best to never check any of those into the repo.
- Added specificity to the Devise Oauth setup to get a bit more information and provide the right callback URL
- Set some additional variables on the `User` model based on the auth hash we get back (e.g. photo, facebook UID, etc).
- Cleaned up the view a bit in terms of styling – mostly just adding spacing / emphasis and removing unneeded styles (e.g. we shouldn't override the btn classes, we just needed to reorder how we load the `bootstrap-social` sass file).
- Opted to use browser validations of `required: true` on form elements as opposed to the `client_side_validations` library – the latter had a bit of weirdness that it would validate even if I didn't try to submit but navigated away. Additionally, since we're using devise for email validation, the client_side_validation doesn't add email validation since it's not directly in the User model.
- [Needs some more work] added option to opt to use facebook photo or switch it out (this is in `_avatars.html.erb`) to other options. Should be fairly straightforward to finish this up though.

**Changes we can think about / discuss**
- Removed the Role needed when signing up – a side note about that is that we had added it when this was still RedesignChallenge and we always hated how it slowed down the sign up process. Since we made the form longer and now have it going to FB as well, I'd rather have that form stay as short as possible (e.g. increase conversions as much as possible), and instead have a "complete your profile" view that they get sent to upon sign in. Removed the custom JS to require the role when signing up.
- Following from the previous point, override the `after_sign_in_path_for` from devise to send to `edit_user_registration_path(setting: 'onboard')` if they have an incomplete profile (e.g. role is not filled out).
- Add in parsley.js validations (basically hooks into the required, format, etc) and is slightly more robust than client_side_validations (https://github.com/guillaumepotier/Parsley.js)
- I know it probably wasn't a "requirement" at the moment, but I'd really want to consider / think about adding the necessary setup here for multiple OAuth providers. In working on this project for a while, I know that we'd potentially want to add it in later down the road, but the current structure / architecture doesn't allow for that. Here's some ways to do it that shouldn't be too hard in terms of setting up the models: http://sourcey.com/rails-4-omniauth-using-devise-with-twitter-facebook-and-linkedin/ + https://github.com/omniauth/omniauth/wiki/Managing-Multiple-Providers. Essentially, you'd add an `Identity` model that belongs to a user and move all the Oauth stuff to that model instead of on the user – for now we still would only support FB, but it just gives us the flexibility to add other OAuth in the future ;-)

**Comment by krnjn** [comment]
Hey @stedmanblake in thinking it over, I think what we've had in terms of implementation from your branch #27 is probably fine as is. I've trimmed it down quite a bit and left this PR as mostly a functionality branch and providing some views. It's mostly from where you started / got us to, but just focused on actually the OAuth flow. Re the points above, most of them hold except that:
- We can use the client_side_validations, just will take some tweaking to get it right, but happy to give it a shot. If it's too much of a pain to get the styles right etc (see here for an open issue: https://github.com/DavyJonesLocker/client_side_validations/issues/617), then we can switch to parsley, etc.
- Adding additional Oauth methods may be kind of a pain and I actually think there is probably better ways to do it than introducing a new model (using an hstore column for example). We can wait until that actually becomes a requirement – want to be sure to respect the time / scope you had for this project.

Other things that I was able to take care of on this branch / already merged in:
- Redirect the user to complete their profile upon sign in/up - this allows us to remove the check for completing the role on sign up.
- Sets the avatar from FB if they are signed in with FB.

Open items:
- Tests! We'll need to update them, but shouldn't be too bad.
- Handling edge cases of sign ups. One big remaining issue is how we handle someone signing up with FB and then wanting to sign in via email, or rather if there is an email address on file already, do we associate that user with the FB login email? This can be kind of a pain and some of those links above should help, but we should try to think through the edge cases with it.

File: app/controllers/omniauth_callbacks_controller.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+
+  def facebook
+    if request.env["omniauth.auth"].info.email.blank?
+      redirect_to "/users/auth/facebook?auth_type=rerequest&scope=email,public_profile"
+    else
```
Comment (krnjn) [inline app/controllers/omniauth_callbacks_controller.rb]: @stedmanblake Here's where I think we'd need the most work / updates on this stream of work – the main thing we'd want to handle is the case where a user exists with an email in our DB and we're trying to sync the accounts – you can see the example code in this post here that should give some tips: https://github.com/omniauth/omniauth/wiki/Managing-Multiple-Providers

Don't worry about finding an `Identity`, you can just try to lookup a `User` first with the given email and then do the associating with the FB login (otherwise the `first_or_create` action with the `.from_omniauth` method is going to create a user with the given oauth params, even if the email is the same, so we'd get an error since our `User` model can't have a user with the same exact email address).

As a bonus, we can think about adding some logic for handling how a currently signed in user would associate their account with the FB login (think of a button inside your settings that simply says something like "Connect with Facebook"). If they click on that they should also update their user record with all of the OAuth info so that they can sign in with FB in the future and have their photo associated etc. This would probably require writing a new method on the User called `update_from_omniauth` so as to ensure that we don't override their name and other settings they may have already set when they signed up.

File: app/views/devise/registrations/forms/_new.html.erb [hunk 2]
```diff
@@ -16,54 +16,26 @@
 
   <div class='row row-fields'>
     <div id='email' class='col-sm-8 col-sm-offset-2'>
-      <div class='form-group'>
```
Comment (krnjn) [inline app/views/devise/registrations/forms/_new.html.erb]: FYI – this is what this page looks like now:

screen shot 2016-09-04 at 9 24 36 am

File: app/views/devise/sessions/_form.html.erb [hunk 3]
```diff
@@ -11,32 +11,15 @@
     <div id='email' class='col-sm-8 col-sm-offset-2'>
       <div class='form-group'>
         <%= f.label :email %>
-        <%= f.email_field :email, autofocus: true, class: 'form-control input-lg' %>
+        <%= f.email_field :email, autofocus: true, required: true, class: 'form-control input-lg' %>
```
Comment (krnjn) [inline app/views/devise/sessions/_form.html.erb]: Updates to this view as well:

**Comment by krnjn** [comment]
Updates to the views for sign in / up:

screen shot 2016-09-04 at 9 24 43 am
screen shot 2016-09-04 at 9 24 36 am

File: app/controllers/omniauth_callbacks_controller.rb [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+      if user.nil?
+        user = User.find_by(email: auth.info.email)
+
+        if user.nil?
+          user = User.create_from_omniauth(auth)
+        elsif user.uid != auth.uid
+          user.update_from_omniauth(auth)
+        end
+      end
```
Comment (krnjn) [inline app/controllers/omniauth_callbacks_controller.rb:23]: @stedmanblake I cleaned up this controller a bit to keep it a bit simpler. The main things I'm checking for here are:
1. An OAuth user with the provider / uid – since they likely hit the "Sign In with FB" link, would like to look up that specific case first. E.g. we only issue on query to sign someone in who's already got an OAuth connected account.
2. If there is no OAuth user, then issue a second query to find an email user – if this email user exists, update their record with the appropriate provider/uid.
3. If there is both no OAuth user nor an email user, issue the third query and create the account.

Comment (krnjn) [inline app/controllers/omniauth_callbacks_controller.rb:23]: One other thing to note here is that I ripped out the case for when a user is signed in – just as an initial stab at keeping it cleaner / simpler for now. The primary reason for that is that I'd want to sketch out those edge cases a bit more. For example what happens in the following:
- User is signed in, tries to connect with FB account, but that FB account (e.g. uid) is connected to a different user account – we shouldn't call `user.update_with_oauth` then, since we'd try to add the same uid to two accounts.
- User is signed in and we find another user account (e.g. not `current_user`) that has the same email as the FB login they tried to use (think that's what you outlined here: https://github.com/Sevenzo/challenge-platform/blob/cbdc18a2660232ae9f8eb064b4a74dae21c66219/app/controllers/omniauth_callbacks_controller.rb#L16), don't think we could just call `update_with_oauth` either as we effectively have a different account with their FB email as we do with their `current_user` and that may cause some weirdness down the line.

In general, those cases are slightly harder to map out, and mostly wanted to make sure we took care of this initial simpler cases first (and add tests to handle all of this as well...). We don't need to worry too much about the feature spec, the request spec should be sufficient that you had started.

Not sure what the exact scope of this feature was, but I think providing a way for only non-signed in users to have the FB flow is sufficient to complete it at this stage as we think more on those other items when you're signed in (keep in mind we'd need to add an interface for that flow which we don't have at the moment either).

**Comment by krnjn** [comment]
Updated layouts post conversation with Masa to get social logins with equal prominence to the regular sign in / up:

screen shot 2016-09-08 at 10 45 36 pm
screen shot 2016-09-08 at 10 45 30 pm

**Comment by stedmanblake** [comment]
@krnjn wow this looks great! Thanks for your notes. I only just now saw them, so had continued to write tests for the edge cases. But that all makes sense re: already-signed-in user flow. I've pushed up those tests in another PR in case we want to cherry pick any of it: #38. 

Other logistics:
1. Mike should be getting us the finalized landing page redesign soon. We'd planned on Thursday (September 8) but I haven't been able to get in touch with him. I'll ask Masa to ping him today.
2. I'll be AFK for most of the weekend, but if anything comes up, feel free to text me and I'll be able to hop on the computer with a few hours' notice.
